// braid-client/dist/UrlParse.js
/*
 * Copyright 2018 Royal Bank of Scotland
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

"use strict";Object.defineProperty(exports, "__esModule", { value: true });function _classCallCheck(instance, Constructor) {if (!(instance instanceof Constructor)) {throw new TypeError("Cannot call a class as a function");}}

function buildURLClass() {
  if (typeof document !== 'undefined' && typeof window !== 'undefined') {
    return BrowserURL;
  } else {var _require =
    require('url'),URL = _require.URL;
    return URL;
  }
}var

BrowserURL =
function BrowserURL(url) {_classCallCheck(this, BrowserURL);
  var parser = document.createElement('a');
  parser.href = url;
  this.protocol = parser.protocol;
  this.host = parser.hostname + ":" + parser.port;
  this.hostname = parser.hostname;
  this.port = parser.port;
};


var url = buildURLClass();exports.default =
url;
